character-sfx.js

```diff
@@ -342,8 +351,50 @@ class CharacterSfx {
   
     };
     _handleNarutoRun();
-    
 
+    // combo
+    const dispatchComboSoundEvent = (soundIndex) =>{
+      this.dispatchEvent(new MessageEvent('meleewhoosh', {
+        data: {
+          index: soundIndex
+        },
+      }));
+    }
+    const _handleCombo = () => {
+      if (this.player.hasAction('use') && this.player.getAction('use').behavior === 'sword') {
+        const comboAnimationName = this.player.getAction('use').animationCombo ?
+          aimAnimations[this.player.getAction('use').animationCombo[this.player.avatar.useAnimationIndex]]
+          : null;
+        if (comboAnimationName) {
+          if (comboAnimationName !== this.lastSwordComboName) {
+            this.swordComboStartTime = timeSeconds;
+            this.alreadyPlayComboSound = false;
+          }
+          const animations = Avatar.getAnimations();
+          const animation = animations.find(a => a.name === comboAnimationName);
+          const animationComboIndices = Avatar.getanimationComboIndices();
+          const animationIndices = animationComboIndices.find(i => i.name === comboAnimationName);
+          const handDeltas = this.player.getAction('use').boneAttachment === 'leftHand' ? animationIndices.rightHandDeltas : animationIndices.leftHandDeltas;
+          const maxDeltaIndex = this.player.getAction('use').boneAttachment === 'leftHand' ? animationIndices.maxRightDeltaIndex : animationIndices.maxLeftDeltaIndex;
+          
+          const ratio = (timeSeconds - this.swordComboStartTime) / animation.duration;
+          if (ratio <= 1 && !this.alreadyPlayComboSound) {
+            const index = Math.floor(ratio * handDeltas.length);
+            if (index > maxDeltaIndex) {
+              this.alreadyPlayComboSound = true;
+              this.playGrunt('attack');
+              const soundIndex = this.player.avatar.useAnimationIndex;
+              dispatchComboSoundEvent(soundIndex);
+            }
+          }
+        }
+        this.lastSwordComboName = comboAnimationName;
+      }
+      
+    };
+
+    _handleCombo();

```

Comment:
This is poor quality unstructured code. The rest of the file has the exact same problem of being random unstructured `if` statements with a huge state vector, but it's gotten to the point where it is difficult for anyone to review this or even understand what is going on here.

IMO this file needs a complete redesign under some sort of structured coding theory before anyone can review it.